src/Generating/FinishGen.cpp

```diff
@@ -666,29 +666,72 @@ void cFinishGenVines::GenFinish(cChunkDesc & a_ChunkDesc)
 
 
 
-
-
 ////////////////////////////////////////////////////////////////////////////////
 // cFinishGenSprinkleFoliage:
 
-bool cFinishGenSprinkleFoliage::TryAddSugarcane(cChunkDesc & a_ChunkDesc, int a_RelX, int a_RelY, int a_RelZ)
+bool cFinishGenSprinkleFoliage::TryAddCactus(cChunkDesc & a_ChunkDesc, int a_RelX, HEIGHTTYPE & a_RelY, int a_RelZ)
 {
+	const int CactusHeight = 1 + (m_Noise.IntNoise2DInt(a_RelX, a_RelZ) % m_MaxCactusHeight);
+
+	if (!IsDesertVariant(a_ChunkDesc.GetBiome(a_RelX, a_RelZ)))
+	{
+		return false;
+	}
+
+	if (a_ChunkDesc.GetBlockType(a_RelX, a_RelY, a_RelZ) != E_BLOCK_SAND)
+	{
+		return false;
+	}
+
 	// We'll be doing comparison to neighbors, so require the coords to be 1 block away from the chunk edges:
 	if (
-		(a_RelX < 1) || (a_RelX >= cChunkDef::Width  - 1) ||
-		(a_RelY < 1) || (a_RelY >= cChunkDef::Height - 2) ||
-		(a_RelZ < 1) || (a_RelZ >= cChunkDef::Width  - 1)
-	)
+			(a_RelX < 1) || (a_RelX >= cChunkDef::Width - 1) ||
+			(a_RelZ < 1) || (a_RelZ >= cChunkDef::Width - 1) ||
+			(a_RelY + CactusHeight >= cChunkDef::Height - 1)
+			)
 	{
 		return false;
 	}
 
-	// Only allow dirt, grass or sand below sugarcane:
+	for (int i = 0; i < CactusHeight; i++)
+	{
+		const bool cactusExists = i != 0;
+
+		const int y = a_RelY + 1;
+		if (
+			cBlockInfo::IsSolid(a_ChunkDesc.GetBlockType(a_RelX + 1, y, a_RelZ)) 	 ||
+			cBlockInfo::IsSolid(a_ChunkDesc.GetBlockType(a_RelX - 1, y, a_RelZ)) 	 ||
+			cBlockInfo::IsSolid(a_ChunkDesc.GetBlockType(a_RelX, 	 y, a_RelZ + 1)) ||
+			cBlockInfo::IsSolid(a_ChunkDesc.GetBlockType(a_RelX, 	 y, a_RelZ - 1))
+		)
+		{
+			return cactusExists;
+		}
+
+		// All conditions are met, we can place a cactus here
+		a_ChunkDesc.SetBlockType(a_RelX, ++a_RelY, a_RelZ, E_BLOCK_CACTUS);
+	}
+
+	// Update chunk height after adding cactus
+	a_ChunkDesc.SetHeight(a_RelX, a_RelZ, a_RelY);
```

Comment:
To be clear, the reason I first suggested having `SetHeight` in these functions was to avoid needing an 'in-out' param. It would have also need some reworking of the logic in `GenFinish` to not use the `Top` variable.  As it is currently, this call is redundant and in fact for `TryAddCactus`, it doesn't even update the heightmap when it bails out early.

Reply:
Yup, I intended to remove them in the new commit, but forgot. I'll remove them.